File: scripts/math_lycee_keyinsights_test.mjs
#!/usr/bin/env node
// Test math lycée keyInsights generation for 3 sample files.
// Uses /api/ai/extract (internal) for PDF text + AI metadata
// Then calls OpenAI gpt-4o-mini directly for the keyInsights format

import { PrismaClient } from '@prisma/client';
import OpenAI from 'openai';
import { config } from 'dotenv';
import fs from 'fs';

config({ path: '/workspace/edutunisie/.env.local' });

const p = new PrismaClient();
const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

const TEST_FILES = [12483, 12234, 12620];
const INTERNAL_TOKEN = process.env.INTERNAL_BULK_TOKEN || 'devmanet-bulk-2026';
const BASE_URL = 'https://examanet.com';

async function getFileText(fileKey) {
  // Use the internal blob API
  const res = await fetch(`${BASE_URL}/api/blob-teacher/${fileKey}`, {
    headers: { 'X-Internal-Token': INTERNAL_TOKEN },
  });
  if (!res.ok) return null;
  const buf = Buffer.from(await res.arrayBuffer());
  // Use a small Python script to extract text
  const tmpPdf = `/tmp/test_${Date.now()}.pdf`;
  const tmpTxt = `/tmp/test_${Date.now()}.txt`;
  fs.writeFileSync(tmpPdf, buf);
  // Use pdftotext if available, else fallback
  try {
    const { execSync } = await import('child_process');
    execSync(`pdftotext -layout "${tmpPdf}" "${tmpTxt}" 2>/dev/null`);
    const text = fs.readFileSync(tmpTxt, 'utf8');
    fs.unlinkSync(tmpPdf);
    fs.unlinkSync(tmpTxt);
    return text.length > 200 ? text : null;
  } catch (e) {
    fs.unlinkSync(tmpPdf);
    if (fs.existsSync(tmpTxt)) fs.unlinkSync(tmpTxt);
    return null;
  }
}

async function extractExercisesMath(num, title, text) {
  if (!text || text.length < 100) return null;
  const nonce = `${num}-${Date.now()}`;
  const system = `Tu es un expert en mathématiques du système éducatif tunisien (lycée 1ère à 4ème année).
Analyse ce document (titre: ${title.slice(0, 120)}) et extrais TOUS les exercices ou parties d'exercices.
Pour CHAQUE exercice/partie: "Exercice N: [sujet/thème court, 5-10 mots] - [résumé FR, 10-20 mots]"
Format strict: commence par "Exercice" puis numéro, puis ":", puis sujet après ":", puis " - " puis résumé.
Exemples valides:
  "Exercice 1: Étude d'une fonction logarithme - Domaine, dérivabilité, limites, tableau de variations."
  "Exercice 2: Géométrie dans l'espace - Calcul de volumes, aires, distances entre points."
  "Exercice 3: Probabilités conditionnelles - Arbre pondéré, événements indépendants, formule de Bayes."
Si le document ne contient pas d'exercices, retourne un JSON vide.
Retourne UNIQUEMENT JSON: {"exercises": ["Exercice 1: ... - ...", ...]}
Limite: 3-12 exercices max (les plus importants).
Nonce: ${nonce}`;

  const resp = await openai.chat.completions.create({
    model: 'gpt-4o-mini',
    messages: [
      { role: 'system', content: system },
      { role: 'user', content: `---DOC---\n${text.slice(0, 25000)}\n---END---` },
    ],
    response_format: { type: 'json_object' },
    temperature: 0.1,
    max_tokens: 3000,
  });
  const parsed = JSON.parse(resp.choices[0].message.content);
  const ex = parsed.exercises || [];
  return ex.filter(e => e.includes('Exercice') && e.includes(':') && e.includes(' - ') && e.length < 350);
}

async function extractSectionsCourse(num, title, text) {
  if (!text || text.length < 100) return null;
  const nonce = `${num}-${Date.now()}`;
  const system = `Tu es un expert en mathématiques du système éducatif tunisien (lycée 1ère à 4ème année).
Analyse ce COURS (titre: ${title.slice(0, 120)}) et extrais les sous-titres / parties principales du cours.
Pour CHAQUE sous-titre ou partie identifiée dans le document:
  "Nom exact du sous-titre tel qu'il apparaît dans le document: concept mathématique clé résumé en 10-20 mots"
Format strict: PAS de parenthèses, PAS de "Section N" au début, PAS de numérotation. Directement le titre du sous-titre, suivi de ":", puis le résumé.
Exemples valides:
  "Introduction à la statistique: Définition, population, individu, effectif, caractère."
  "Vocabulaire en statistiques: Caractères, classes, effectifs, fréquences et pourcentages."
  "Représentations graphiques: Diagrammes à barres, polygones, histogrammes, circulaires."
  "Paramètres d'un caractère statistique: Position et dispersion, mode, moyenne, médiane."
Si le document ne contient pas de sous-titres clairs, retourne un JSON vide.
Retourne UNIQUEMENT JSON: {"sections": ["Titre du sous-titre: résumé", ...]}
Limite: 3-10 sections max (les plus importantes, dans l'ordre du document).
Nonce: ${nonce}`;

  const resp = await openai.chat.completions.create({
    model: 'gpt-4o-mini',
    messages: [
      { role: 'system', content: system },
      { role: 'user', content: `---DOC---\n${text.slice(0, 25000)}\n---END---` },
    ],
    response_format: { type: 'json_object' },
    temperature: 0.1,
    max_tokens: 3000,
  });
  const parsed = JSON.parse(resp.choices[0].message.content);
  const sec = parsed.sections || [];
  // Accept any line that has a colon, no parenthesis, and reasonable length
  return sec.filter(s => !s.startsWith('(') && s.includes(':') && s.length < 350);
}

async function main() {
  // Use findMany to avoid numericId type issue
  const records = await p.resource.findMany({
    where: {
      numericId: { in: TEST_FILES },
      subject: { slug: 'mathematiques' },
      class: { slug: { in: ['1ere-secondaire', '2eme-secondaire', '3eme-secondaire', '4eme-secondaire'] } },
      status: 'PUBLISHED',
    },
    include: { class: true, section: true },
  });
  const files = records.map(r => ({
    id: r.id,
    numericId: r.numericId,
    title: r.title,
    type: r.type,
    fileKey: r.fileKey,
    fileUrl: r.fileUrl,
    class_name: r.class?.nameFr,
    section_name: r.section?.nameFr,
  })).sort((a, b) => a.numericId - b.numericId);
  
  const results = [];
  for (const f of files) {
    console.log('\n' + '='.repeat(70));
    console.log(`📄 #${f.numericId} (${f.type})`);
    console.log(`   ${f.title}`);
    console.log(`   ${f.class_name} / ${f.section_name || '(no section)'}`);
    console.log('='.repeat(70));
    
    let text;
    try {
      text = await getFileText(f.fileKey);
    } catch (e) {
      console.log(`   ❌ Download failed: ${e.message}`);
      results.push({ ...f, error: e.message });
      continue;
    }
    
    if (!text) {
      console.log(`   ❌ No text extracted (PDF may be scanned)`);
      results.push({ ...f, error: 'no text' });
      continue;
    }
    
    console.log(`   Text: ${text.length} chars`);
    console.log(`   Preview: ${text.slice(0, 200).replace(/\n/g, ' ')}...`);
    
    let insights;
    try {
      insights = f.type === 'COURSE' 
        ? await extractSectionsCourse(f.numericId, f.title, text)
        : await extractExercisesMath(f.numericId, f.title, text);
    } catch (e) {
      console.log(`   ❌ API error: ${e.message}`);
      results.push({ ...f, error: 'api: ' + e.message });
      continue;
    }
    
    if (!insights || insights.length === 0) {
      console.log(`   ⚠️ No insights returned`);
      results.push({ ...f, keyInsights: [], text_len: text.length });
      continue;
    }
    
    console.log(`\n   📋 Generated ${insights.length} keyInsights:`);
    insights.forEach((ki, i) => console.log(`      ${i + 1}. ${ki}`));
    
    results.push({ ...f, keyInsights: insights, text_len: text.length });
    await new Promise(r => setTimeout(r, 1000));
  }
  
  console.log('\n' + '='.repeat(70));
  console.log('✅ Tests complete.');
  console.log(`   Files tested: ${results.length}`);
  console.log(`   Files with keyInsights: ${results.filter(r => r.keyInsights?.length).length}`);
  
  fs.writeFileSync('/tmp/math_lycee_test_results.json', JSON.stringify(results, null, 2));
  console.log(`   Results saved to /tmp/math_lycee_test_results.json`);
  
  await p.$disconnect();
}

main().catch(e => { console.error(e); process.exit(1); });

File: scripts/math_lycee_keyinsights_test.py
#!/usr/bin/env python3
"""Test math lycée keyInsights generation for 3 sample files (1 série, 1 cours, 1 devoir).

Same approach as physique lycée:
- DEVOIR/EXERCISE: "Exercice N: [topic] - [summary 15-25 words]"
- COURSE: "Section N (Title): [concept summary]" - one per subtitle

For math, we use "Exercice" / "Exercice de cours" / "Partie" etc. as appropriate.
"""
import os
import json
import time
import requests
import tempfile
from pypdf import PdfReader
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv('/workspace/edutunisie/.env.local')

INTERNAL_TOKEN = os.environ.get("INTERNAL_BULK_TOKEN", "devmanet-bulk-2026")

# Test files: (numericId, type, expected_format)
TEST_FILES = [
    3818,  # EXERCISE: Série d'exercices Barycentre
    4077,  # COURSE: Pollution des océans
    93,    # DEVOIR: Dérivabilité et limites
]

client = OpenAI(api_key=os.environ['OPENAI_API_KEY'])

def get_blob_text(file_url):
    """Download via Vercel Blob URL, extract text with PyPDF."""
    try:
        r = requests.get(file_url, timeout=30)
        if r.status_code != 200:
            return None, f'HTTP {r.status_code}'
        data = r.content
    except Exception as e:
        return None, f'download: {e}'
    
    if not data or len(data) < 100:
        return None, 'empty blob'
    
    with tempfile.NamedTemporaryFile(suffix='.pdf', delete=False) as tmp:
        tmp.write(data)
        tmp_path = tmp.name
    
    try:
        reader = PdfReader(tmp_path)
        text = '\n'.join(page.extract_text() or '' for page in reader.pages)
        os.unlink(tmp_path)
        if len(text) > 200:
            return text, 'pypdf'
        return None, 'too little text'
    except Exception as e:
        os.unlink(tmp_path)
        return None, f'pypdf: {e}'


def extract_exercises_math(num, title, text):
    """For DEVOIR or EXERCISE: extract each exercise as 'Exercice N: topic - summary'."""
    if not text or len(text) < 100:
        return None
    
    nonce = f'{num}-{time.time()}'
    system = f"""Tu es un expert en mathématiques du système éducatif tunisien (lycée 1ère à 4ème année).
Analyse ce document (titre: {title[:120]}) et extrais TOUS les exercices ou parties d'exercices.
Pour CHAQUE exercice/partie: "Exercice N: [sujet/thème court, 5-10 mots] - [résumé FR, 10-20 mots]"
Format strict: commence par "Exercice" puis numéro, puis ":", puis sujet après ":", puis " - " puis résumé.
Exemples valides:
  "Exercice 1: Étude d'une fonction logarithme - Domaine, dérivabilité, limites, tableau de variations."
  "Exercice 2: Géométrie dans l'espace - Calcul de volumes, aires, distances entre points."
  "Exercice 3: Probabilités conditionnelles - Arbre pondéré, événements indépendants, formule de Bayes."
Si le document ne contient pas d'exercices, retourne un JSON vide.
Retourne UNIQUEMENT JSON: {{"exercises": ["Exercice 1: ... - ...", ...]}}
Limite: 3-12 exercices max (les plus importants).
Nonce: {nonce}"""
    
    try:
        resp = client.chat.completions.create(
            model='gpt-4o-mini',
            messages=[
                {'role': 'system', 'content': system},
                {'role': 'user', 'content': f'---DOC---\n{text[:25000]}\n---END---'},
            ],
            response_format={'type': 'json_object'},
            temperature=0.1,
            max_tokens=3000,
        )
        parsed = json.loads(resp.choices[0].message.content)
        ex = parsed.get('exercises', [])
        valid = [e for e in ex if 'Exercice' in e and ':' in e and ' - ' in e and len(e) < 350]
        return valid if valid else None
    except Exception as e:
        return {'error': str(e)}


def extract_sections_course(num, title, text):
    """For COURSE: extract each section/subsection as 'Section N (Title): summary'."""
    if not text or len(text) < 100:
        return None
    
    nonce = f'{num}-{time.time()}'
    system = f"""Tu es un expert en mathématiques du système éducatif tunisien (lycée 1ère à 4ème année).
Analyse ce COURS (titre: {title[:120]}) et extrais les sous-titres / parties principales du cours.
Pour CHAQUE sous-titre ou partie identifiée dans le document: 
  "Section [N] (Titre du sous-titre): [concept mathématique clé, 10-20 mots]"
Format strict: commence par "Section" puis numéro, puis " (Titre)", puis ":", puis concept résumé.
Exemples valides:
  "Section 1 (Définition et notations): Introduction aux barycentres de 2 puis 3 points pondérés."
  "Section 2 (Propriétés des barycentres): Linéarité, associativité, cas particuliers du centre de gravité."
  "Section 3 (Applications géométriques): Construction du barycentre, alignement, coordonnées."
Si le document ne contient pas de sous-titres clairs, retourne un JSON vide.
Retourne UNIQUEMENT JSON: {{"sections": ["Section 1 (Titre): ...", ...]}}
Limite: 3-10 sections max (les plus importantes).
Nonce: {nonce}"""
    
    try:
        resp = client.chat.completions.create(
            model='gpt-4o-mini',
            messages=[
                {'role': 'system', 'content': system},
                {'role': 'user', 'content': f'---DOC---\n{text[:25000]}\n---END---'},
            ],
            response_format={'type': 'json_object'},
            temperature=0.1,
            max_tokens=3000,
        )
        parsed = json.loads(resp.choices[0].message.content)
        sec = parsed.get('sections', [])
        valid = [s for s in sec if 'Section' in s and ':' in s and len(s) < 350]
        return valid if valid else None
    except Exception as e:
        return {'error': str(e)}


def main():
    # Get test file details
    from prisma import Prisma
    p = Prisma()
    p.connect()
    
    files = []
    for numid in TEST_FILES:
        r = p.resource.find_first(
            where={'numericId': numid},
            include={'class': True, 'section': True}
        )
        if r:
            data = r.model_dump() if hasattr(r, 'model_dump') else dict(r)
            cls_obj = data.get('class')
            sec_obj = data.get('section')
            cls_name = cls_obj.get('nameFr') if cls_obj else None
            sec_name = sec_obj.get('nameFr') if sec_obj else None
            files.append({
                'id': r.id, 'numericId': r.numericId, 'title': r.title,
                'type': r.type, 'class': cls_name, 'section': sec_name,
                'fileUrl': r.fileUrl,
            })
    p.disconnect()
    
    results = []
    for f in files:
        print(f"\n{'='*70}")
        print(f"📄 #{f['numericId']} ({f['type']})")
        print(f"   {f['title']}")
        print(f"   {f['class']} / {f['section']}")
        print(f"{'='*70}")
        
        text, source = get_blob_text(f['fileUrl'])
        if not text:
            print(f"   ❌ Could not extract text: {source}")
            results.append({**f, 'error': source})
            continue
        
        print(f"   Text: {len(text)} chars (via {source})")
        
        # Choose extractor based on type
        if f['type'] == 'COURSE':
            insights = extract_sections_course(f['numericId'], f['title'], text)
        else:
            insights = extract_exercises_math(f['numericId'], f['title'], text)
        
        if not insights:
            print(f"   ⚠️ No insights returned")
            results.append({**f, 'keyInsights': [], 'text_len': len(text)})
            continue
        
        if isinstance(insights, dict) and 'error' in insights:
            print(f"   ❌ API error: {insights['error']}")
            results.append({**f, 'error': insights['error']})
            continue
        
        print(f"\n   📋 Generated {len(insights)} keyInsights:")
        for i, ki in enumerate(insights, 1):
            print(f"      {i}. {ki}")
        
        results.append({**f, 'keyInsights': insights, 'text_len': len(text), 'method': source})
        time.sleep(1)  # rate limit
    
    # Save results for user review
    with open('/tmp/math_lycee_test_results.json', 'w') as fp:
        json.dump(results, fp, ensure_ascii=False, indent=2)
    
    print(f"\n{'='*70}")
    print(f"✅ Tests complete. Results saved to /tmp/math_lycee_test_results.json")
    print(f"   Files tested: {len(results)}")
    print(f"   Files with keyInsights: {sum(1 for r in results if r.get('keyInsights'))}")
    return results


if __name__ == '__main__':
    main()
